libbpf-tools/biostacks.bpf.c

```diff
@@ -0,0 +1,115 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2020 Wenbo Zhang
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include "biostacks.h"
+#include "bits.bpf.h"
+#include "maps.bpf.h"
+
+#define MAX_ENTRIES	10240
+#define NULL		0
+
+const volatile bool targ_ms = false;
+const volatile dev_t targ_dev = -1;
+
+struct internal_rqinfo {
+	u64 start_ts;
+	struct rqinfo rqinfo;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct request *);
+	__type(value, struct internal_rqinfo);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} rqinfos SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, struct rqinfo);
+	__type(value, struct hist);
+	__uint(map_flags, BPF_F_NO_PREALLOC);
+} hists SEC(".maps");
+
+static struct hist zero;
+
+static __always_inline
+int trace_start(void *ctx, struct request *rq, bool merge_bio)
+{
+	struct internal_rqinfo *i_rqinfop = NULL, i_rqinfo = {};
+	struct gendisk *disk = BPF_CORE_READ(rq, rq_disk);
+	dev_t dev;
+
+	dev = disk ? MKDEV(BPF_CORE_READ(disk, major),
+			BPF_CORE_READ(disk, first_minor)) : 0;
```

Comment:
I am intentional, I think as we have `partitions__get_by_dev`, so I changed 

```c
struct rqinfo {
 	__u32 pid;
 	int kern_stack_size;
 	__u64 kern_stack[MAX_STACK];
 	char comm[TASK_COMM_LEN];
 	__u32 dev;
 };
```

last member from `char disk[32] to _u32 dev`;

I think it's cheaper than calling `bpf_probe_read_kernel_str` and `BPF_CORE_READ_STR_INTO`, am I right?

Reply:
I meant that you are doing those BPF_CORE_READ()s even if you don't need to filter by device, so I suggested to read that only if targ_dev != -1:

```
if (targ_dev != -1) {
    struct gendisk *disk = BPF_CORE_READ(rq, rq_disk);
    dev_t dev;
    
    dev = disk ? MKDEV(BPF_CORE_READ(disk, major), BPF_CORE_READ(disk, first_minor)) : 0;
    if (targ_dev != dev)
        return 0;
}
```
?

Reply:
Yes I knew what you mean,  but I want to print disk name before printing call stack:

comm pid **disk_name**     # eg:     etcd           12849  **sda**
kernel stack
hist

So I always need `dev`, and in userspace call 

```c
part = partitions__get_by_dev(parts, next_key.dev);
printf("%-14.14s %-6d %-7s\n", next_key.comm, next_key.pid, part ? part->name : "Unknown");
```

Before walking kernel stacks.

> I think it's cheaper than calling bpf_probe_read_kernel_str and BPF_CORE_READ_STR_INTO, am I right?

I meant at last, I read & save the disk name in the kernel by calling `bpf_probe_read_kernel_str` and `BPF_CORE_READ_STR_INTO`  every time,  I think it may be more expensive than just read & save dev. Am I right?

Reply:
ah, I totally missed dev use at line 64, sorry, never mind then

re: overhead, tbh, I don't know what's more or less expensive. I'm just always trying to avoid string comparisons on BPF side, so I prefer dev_t. But then a bunch of user-space complexity could have been saved by just having dev_name. So it's not clear what's the best solution. Both would be fine, I think.

Reply:
Get it, thanks for sharing your experience. :)